src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -42,23 +43,67 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
   }
 };
 
+// Reserve space for a flush token, so 'push_back(Token)' always succeeds.
+AsyncLogWriter::Buffer::Buffer(size_t capacity) : _pos(0) {
+  assert(capacity > (AsyncLogWriter::Token.size()), "AsyncLogBuffer is too small!");
+  _buf = NEW_C_HEAP_ARRAY_RETURN_NULL(char, capacity, mtLogging);
+  _capacity = capacity - (Token.size());
+}
+
+AsyncLogWriter::Buffer::~Buffer() {
+  FREE_C_HEAP_ARRAY(char, _buf);
+}
+
+bool AsyncLogWriter::Buffer::is_valid() const {
+  return _capacity > 0 && _buf != nullptr;
+}
+
+bool AsyncLogWriter::Buffer::push_back(const AsyncLogMessage& msg) {
+  size_t sz = msg.size();
+
+  if (_pos + sz <= _capacity || (&msg == &AsyncLogWriter::Token)) {
+    memcpy(_buf + _pos, &msg, sizeof(AsyncLogMessage));
+    char* dest = _buf + _pos + sizeof(AsyncLogMessage);
+    if (msg.message() != nullptr) {
+      strcpy(dest, msg.message());
+    } else {
+      dest = nullptr;
+    }
+    auto p = reinterpret_cast<AsyncLogMessage*>(_buf + _pos);
+    p->set_message(dest);
+    _pos += sz;
+    return true;
```

Comment:
That's pretty hacky... :-) And `set_message` only exists because you need to repair the message pointer here, right?

Proposal:

Make AsyncLogMessage self-contained and variable-sized by letting the message follow the header. Like this:

```
class AsyncLogMessage {
  LogFileStreamOutput* _output;
  const LogDecorations _decorations;
  const size_t _len; // optional, or could also be 32- maybe even 16 bit
public:
  size_t len_in_bytes() const { return align_up(sizeof(AsyncLogMessage) + _len); }
  const char* message() { return (const char*) (this + 1); }
};
```

Then give it a placement new() and place it inside the Buffer right away instead of creating it on the stack and then copying it to the buffer. Just an idea.

Reply:
I prefer trailing 0 because `AsyncLogMessage::message()` can be treated as a valid c-string and pass to LogFileStreamOutput. If so, `len` is optional indeed. The benefit of `len` is that we can iterate buffer faster. We will have O(1) time-complexity instead of O(len) in `Buffer::Iterator::next() `

I see your point. It's more elegant to use `implicit` message. The reason I need to handle `msg= nullptr` is that flushing Token uses it. I can replace it with "" instead.

Reply:
I update to use in-place new for AsyncLogMessage objects. For safety, I change AsyncLogMessage to a nested class and rename it to 'Message'. It's private so it's not accessible by outside.

Previously, we had at least one extra copy of `LogDecorations`. Now it aligns with c-str message. We materialize them only in constructor. Because we have to support asynchronous execution, this is optimal.